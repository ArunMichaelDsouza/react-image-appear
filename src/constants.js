// Global constants

const LOADER = 'data:image/gif;base64,R0lGODlhMgAyAPMAAP////f39+/v7+bm5t7e3tbW1szMzMXFxb29vbW1ta2traWlpZmZmYyMjISEhHNzcyH/C05FVFNDQVBFMi4wAwEAAAAh+QQFCAAQACH+FlJlc2l6ZWQgd2l0aCBlemdpZi5jb20ALAAAAAAyADIAAAX/ICSOZGmKzfM4Z+u+cKqucG2383zvZbIwC1NOZRoQCAPeScFoMhSl4cN0PCaVpJ+TER2WqlVsdgskSUlGMEFMInNH55EayR5pnUFRHJIG19tkCXpeInN/Y1t5cXNrPAgKCQgmd01Qi2onVicHCp2QJm4QZ3NXI310PZ6egnZkCzIqLIamjCYJqqqSImRQMiwQaqWMRyaPuKuSd1Anp4XDxCe3x52CCnkvxKekLwjSx6w32pg23qrgNuLQPOXLPNts0ro8faWHbAP12AX7/P0FjfeeaSLgr+A+gErSkTLIkI1ANQwNOnxYJWLBiRTXWOT3J6O9Gvg+VsmHbmSdWe6CfCUcNo9lyWcth5H8IpDPzC9JFGoqIjCnSWZ+gD0jOVQOGJ+oVDqrRROl0KOnkjhNxzMoCUNYlRr92ZSoyllOjd50IW5pmKdcD71DGTbgOLTqmHZ8CxfhO4xW66JpmzKv3qt0sURtehZwYbcIzaqjhepjpsCObfCNDMMvlhAAIfkEBQgAEAAsAAAAACgAHAAABcogJI5kaYoKwyxn675Qqq5wDc+zDSOJkpg4lanhcDBsCIVSgSgFj6WH9NGo9ZY/0rPkmEqtS2VWtB0xvN9aWKkNkrreKjg8LkMaaIcutj6Q3SJoD3sQSXR/OCJwU3omBAUFBCZXS01lgiePkJAmawoQW4tSjSWbppIjlGIyKiyYJ6axqHyVMa2BcS2asZuSlE0neKMwvLGFYy5FNru8s4Q1xZzPe7zT073W2dolAwTe3+AEA9sk3eHn4uQi6OzqEOzo7vDn7u/z3vUhACH5BAUIABAALAMAAAAtABYAAAXLICSOZFkiipKYbOu+aKq+dG3Ksq2LRFEQN5zitlisdjyfD0gSDkuMKOO5UypLztJCGkVCrMumsMntesEFMY605VKr6FFW1JYuvCM0cw4pM/AjPWAic3VRdzQDBAQDJnoQc34sDg4NIoqLi46DWYYMiCMND6MPf5mnjXlgBDEpK5KhDqSjDhCnt6lfVkAxR2VUsrOkEJi3mY2CPy11iMHCtIHGp7ZML0UQos+kliTFxrk22dq1LdKaO+Kz5DTGeM/cOsd44vCAgA31LCEAIfkEBQgAEAAsCwAAACcAHAAABcogJI6kSBRoqa4se6JpK88iDNPzQBCDasckRCKBoOl2vNKvoFI4FUUZEqn8lRJPZ26aHC2DWe2MSyB9R9hsVEr2WkXpZwIHIfcgZ0hYQYccpzVvcU5zfXZ4b3srCwt8K21fgwqFIwsMlwyUJHYvN2GaCpiYCy5TPZ0iYWuWopgsf3cqCHIirK2XpK+xLUMQtrcMjn0tocAMucMyv6PJdLfCzTPFwdHDCsjVfQwP3N3eDw3ZJNvf5eDiIubq6BDq5uwO7t4O7BDx8vQhACH5BAUIABAALBYAAAAcACcAAAXEIDQQJGSeaKqaI1musOq6cX3OL0oUBWFDOF+KxxPGgkMiz4a8KZe1punZY+Kcz5+U+gNegVztN5xKJBC6K9VoSijeisOpRSNDEG64IoGiT5VGeXpvUUptg3p8NkuCiApoXXeOb4qREI1wlZYmg2ebKQhwkJ+gmqQpCgyqq6wMCpGprbKuXbO2tbayXQu5rAuRvL2/XQ0NpyoND8rGxybKzw4Mx8nP0KfU1c/Mmw7Z2duW3d7Lp+Le094OzRDmD9Lr7A7gIQAh+QQFCAAQACwcAAMAFgAsAAAFwyAkjgNBDGOqrqaJrrDYtjFczkTN4q8u3jNfCpcTjohGGQ5WKBSPPBWh2XxCkEeqdjjrab8pYPFL3l3JWqsUTfVN0UmyWvd2Jrn3FULB7/sVCDF7f4SAMIWIh4iEMAmLfgkxjo+Rhwt5IwoMmwqYm58LlUILn6WXQpqlpjUODRCkqqWdKQ0Ptg4isLEMpyK2vwwjuqopv7cpw5u9EA7GD67IsirODzALC7Mpzca4d9R527/dSdTBSbXc4M6Y6XkNzeY6IQAh+QQFCAAQACwWAAsAHAAnAAAFyCAkjmRpEsRgrqw4oKjazi8Mz62t46e+8y6fTQasCQlAkvGXJPmIzZERGlVSqyVCYcvtFpBArXf8BZLP5vM4re4m29wkIoFdIRR4RJ2E7yf0dQl9g3RYd4OETQsLEIKIfoAtCwyUjI2PfiQMDw8NIpSgCiOOjyObnJ0QoJUkh4kiqKgQk6uWo4ORsZyfqwwrCcAkug8itKC2M8OmvUDKxb3ILM68q6LJuiTGrNexJdq+3LIltTjTI4sM1uG7exAOsQ7tIu8P8fIhACH5BAUIABAALAsAFgAnABwAAAXPICSOZGmSA6GuLDGccAylbe3K+Gjveb7bPdyvFhQOVUUZMskUEQpQQjMJrRakU6HVmlVurVhZo5FLJCDPbzVsajze5BNCQT9D1GvYe09WMBgLInSDCCN4BScOe3B+fwwKEIN1JGlbJ4tvEI6OEAmSCoUkW2wiiotkm3+CnydRbZgPIqkMIp6Sdj2mfLKpI5+QPW6LDiOzI7aDuDKwJMbHn6Exum/Exb0jc7c407HN18+SOMJwJc4k2jgMitXemyYIZtFZ5l0xC5uB9Tj3gPohACH5BAUIABAALAMAHAAsABYAAAXLICSOZGme5DCgrOg8T9OiA2Gv8/jCcG7aQAKO1eAZHb5RLQhEMXZGGDIJWTJvJWj0gVQscgVC9RpcFbc82YLBVqAIhXhBRAYOtMcRe+9GKBQIInJxYnR1aFMia3sMC35/gBCDcSlkDHkljGwQkJCSk4UjTCs7MiWLfJydbp+DJ1gtCpoMIqusk2FUmZqstiO4uiOobF+1q7/AwbMkvsiDoT7DjczHoslJ0iXNznK6sqnUnSaTwV7T2tUkcLnBsentwQmdCfD18gr08CEAIfkEBQgAEAAsAAAWACcAHAAABcgg1DxkaT4MpK5s67LjKaNvbc+4rbf4vP8QR8/kAP6Ew6JxyWwCFwzGwrmERqPUn/UanWZfii1X+nWJx1NE4kcY2BTjsQKSUNgRtQFhT6idr14QdoN4BAUFfRB8e24ucWQrdYMKCYaHiIqLiS1wgC6TdhCXl5mLjS1Wcy2ShKKjBSqam0AIoGuuo7Gapz+sdyqvsLqLRr6UK8ErsrNvoCzJyrs7xrfAryx6mtPOz9csstut3bkt2jtqxy3Q2IxluKTuNuvxL+Q6IQAh+QQFCAAQACwAAAoAHAAoAAAFwiAkjmRpnmTzOAzqmswjr28NOfPs2Ciezw1eSfXTtYQjX5GGFBGXwaZImdtJk7+ryRfVehWMsHjMUAjB5HSZp26z22neAj5eCOd0u3dvSygUCHx+f39eg4R/CVIIh4gKikiNjooEBDYIjogJgQQFnpYukoSQEJ6mlgOVAySipKWmn6mVBKsimKMmsJ4Qs7Osf4ElugWWvZU8nboixqA1w6DMNsmwI9HOytXG19jL2i/DJdYu08Th3i/TJuJa61ft7r4hACH5BAkIABAALAAAAAAyADIAAAXlICSOZGmeaKqubOu+cJwyztPIOMo8vJ3/o1rvAQQ2hr5iDklU4oQ9hxPHZExjx6H0alIwFgoTlHfjjhKM9NfENI8WavWCNE6a4XH5CFm+5+NgEFB9bhB4f2sNhIUih39zjCiOepEpk5UseJCYnCwICqChogoInJ+jqKSYqayrrKiYCa+iCZyys7WdukAFvQS7JQS9w8AiwsPIu8fIyZ3MzxADA5XLzwW/AwTa04XWviPa4dxm3iTh4oXVBSbn2pHDv8Ht47vZ7cXG8/j25/gQ7fGKAQwIDKC/aOfoFeN3sKHDhyJCAAA7', LOADER_STYLE = {
    width: 40,
    height: 40,
    position: 'absolute',
    left: 'calc((100% - 40px) / 2)',
    top: 'calc((100% - 40px) / 2)'
}, PLACEHOLDER = 'data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz4NCjwhLS0gR2VuZXJhdG9yOiBBZG9iZSBJbGx1c3RyYXRvciAxNi4wLjAsIFNWRyBFeHBvcnQgUGx1Zy1JbiAuIFNWRyBWZXJzaW9uOiA2LjAwIEJ1aWxkIDApICAtLT4NCjwhRE9DVFlQRSBzdmcgUFVCTElDICItLy9XM0MvL0RURCBTVkcgMS4xLy9FTiIgImh0dHA6Ly93d3cudzMub3JnL0dyYXBoaWNzL1NWRy8xLjEvRFREL3N2ZzExLmR0ZCI+DQo8c3ZnIHZlcnNpb249IjEuMSIgaWQ9IkxheWVyXzEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4Ig0KCSB3aWR0aD0iNDIuNjVweCIgaGVpZ2h0PSIzMnB4IiB2aWV3Qm94PSIwIDAgNDIuNjUgMzIiIGVuYWJsZS1iYWNrZ3JvdW5kPSJuZXcgMCAwIDQyLjY1IDMyIiB4bWw6c3BhY2U9InByZXNlcnZlIj4NCjxnPg0KCTxnPg0KCQk8cGF0aCBmaWxsPSIjRDNEM0QzIiBkPSJNMCwwaDQyLjYzN2MwLjAyNCwwLjM1NCwwLjAwNCwwLjcxLDAuMDExLDEuMDY2YzAsMTAuMzEzLDAsMjAuNjI0LDAsMzAuOTM0SDBWMHogTTE1Ljg0OSwxMS41MDINCgkJCWMwLDIuOTYzLTAuMDAyLDUuOTI3LDAuMDAxLDguODkxYzMuNjQ0LDAuMDAzLDcuMjksMC4wMDMsMTAuOTM2LDBjMC4wMDItMi45NjQsMC4wMDItNS45MjcsMC04Ljg5MQ0KCQkJQzIzLjE0MSwxMS40OTgsMTkuNDk1LDExLjQ5OCwxNS44NDksMTEuNTAyeiIvPg0KCTwvZz4NCgk8Zz4NCgkJPHBhdGggZmlsbD0iI0QzRDNEMyIgZD0iTTE2LjUzNywxMi4xODZjMy4xODgsMCw2LjM3NSwwLDkuNTY0LDBjMC4wMDEsMi41MDksMC4wMDEsNS4wMTQtMC4wMDIsNy41MjENCgkJCWMtMy4xODgsMC4wMDMtNi4zNzUsMC4wMDMtOS41NjMsMEMxNi41MzMsMTcuMTk5LDE2LjUzMywxNC42OTQsMTYuNTM3LDEyLjE4NnogTTE3LjIxNSwxMi44N2MwLDIuMDUyLTAuMDAyLDQuMTAzLDAuMDAyLDYuMTU1DQoJCQljMC40NTcsMCwwLjkxNSwwLjAwMywxLjM3My0wLjAwNGMwLjQxLTAuNDA5LDAuODIxLTAuODIxLDEuMjMyLTEuMjMyYzAuMDU2LTAuMDU0LDAuMTA2LTAuMTIxLDAuMTgzLTAuMTQ3DQoJCQljMC42NjUtMC4yMjYsMS4zMzUtMC40NDEsMi0wLjY3MmMwLjIyNi0wLjY4MSwwLjQ1MS0xLjM2MSwwLjY3OS0yLjA0MmMwLjE3OSwwLjM1MiwwLjM1NCwwLjcwNywwLjUzMSwxLjA2MQ0KCQkJYzAuMDY2LDAuMTM1LDAuMTQzLDAuMjY2LDAuMTg4LDAuNDA4YzAuMjA3LDAuNjI2LDAuNDE3LDEuMjUxLDAuNjI1LDEuODc4YzAuMDIxLDAuMDc4LDAuMDg3LDAuMTM0LDAuMTQ1LDAuMTkNCgkJCWMwLjE5LDAuMTg1LDAuMzcxLDAuMzc3LDAuNTY1LDAuNTZjMC4yMjgsMCwwLjQ1MywwLjAwMiwwLjY4MywwYzAuMDAyLTIuMDUyLDAuMDAyLTQuMTAzLDAtNi4xNTUNCgkJCUMyMi42ODYsMTIuODY4LDE5Ljk1LDEyLjg2OCwxNy4yMTUsMTIuODd6Ii8+DQoJPC9nPg0KCTxwYXRoIGZpbGw9IiNEM0QzRDMiIGQ9Ik0xOS4zNzcsMTMuNTgyYzAuNDctMC4xMiwwLjk5NywwLjE0OSwxLjE3NSwwLjU5OWMwLjIwMywwLjQ1NSwwLjAxLDEuMDM2LTAuNDI0LDEuMjc5DQoJCWMtMC40MTgsMC4yNjEtMS4wMTYsMC4xNTEtMS4zMTYtMC4yMzljLTAuMzA0LTAuMzY1LTAuMjk4LTAuOTM5LDAuMDEyLTEuMjk5QzE4Ljk2NSwxMy43NTMsMTkuMTYzLDEzLjYzMiwxOS4zNzcsMTMuNTgyeiIvPg0KPC9nPg0KPC9zdmc+DQo=', PLACEHOLDER_STYLE = {
    position: 'relative',
    display: 'inline-block',
    backgroundColor: '#f0f0f0',
    backgroundPosition: 'center center',
    backgroundSize: 'cover',
    backgroundRepeat: 'no-repeat'
}, ANIMATIONS = '@keyframes fadeIn{from{opacity:0}to{opacity:1}}@keyframes fadeInUp{from{opacity:0;transform:translate3d(0,20%,0)}to{opacity:1;transform:translate3d(0,0,0)}}@keyframes fadeInDown{from{opacity:0;transform:translate3d(0,-20%,0)}to{opacity:1;transform:translate3d(0,0,0)}}@keyframes fadeInLeft{from{opacity:0;transform:translate3d(-20%,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}@keyframes fadeInRight{from{opacity:0;transform:translate3d(20%,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}@keyframes bounceIn{from{transform:scale3d(.3,.3,.3)}20%{transform:scale3d(1.1,1.1,1.1)}40%{transform:scale3d(.9,.9,.9)}60%{transform:scale3d(1.03,1.03,1.03)}80%{transform:scale3d(.97,.97,.97)}to{transform:scale3d(1,1,1)}}@keyframes bounceInUp{from{transform:translate3d(0,20%,0)}60%{transform:translate3d(0,-25px,0)}75%{transform:translate3d(0,10px,0)}90%{transform:translate3d(0,-5px,0)}to{transform:translate3d(0,0,0)}}@keyframes bounceInDown{from{transform:translate3d(0,-20%,0)}60%{transform:translate3d(0,25px,0)}75%{transform:translate3d(0,-10px,0)}90%{transform:translate3d(0,5px,0)}to{transform:translate3d(0,0,0)}}@keyframes bounceInLeft{from{transform:translate3d(-20%,0,0)}60%{transform:translate3d(25px,0,0)}75%{transform:translate3d(-10px,0,0)}90%{transform:translate3d(5px,0,0)}to{transform:translate3d(0,0,0)}}@keyframes bounceInRight{from{transform:translate3d(20%,0,0)}60%{transform:translate3d(-25px,0,0)}75%{transform:translate3d(10px,0,0)}90%{transform:translate3d(-5px,0,0)}to{transform:translate3d(0,0,0)}}@keyframes flipInX{from{transform:perspective(400px) rotate3d(1,0,0,90deg);animation-timing-function:ease-in}40%{transform:perspective(400px) rotate3d(1,0,0,-20deg);animation-timing-function:ease-in}60%{transform:perspective(400px) rotate3d(1,0,0,10deg)}80%{transform:perspective(400px) rotate3d(1,0,0,-5deg)}to{transform:perspective(400px)}}@keyframes flipInY{from{transform:perspective(400px) rotate3d(0,1,0,90deg);animation-timing-function:ease-in}40%{transform:perspective(400px) rotate3d(0,1,0,-20deg);animation-timing-function:ease-in}60%{transform:perspective(400px) rotate3d(0,1,0,10deg)}80%{transform:perspective(400px) rotate3d(0,1,0,-5deg)}to{transform:perspective(400px)}}@keyframes zoomIn{from{transform:scale3d(.3,.3,.3)}to{transform:scale3d(1,1,1)}}@keyframes blurIn{from{-webkit-filter:blur(12px);filter:blur(12px)}to{-webkit-filter:blur(0);filter:blur(0)}}@keyframes blurInUp{from{-webkit-filter:blur(15px);filter:blur(15px);transform:translate3d(0,20%,0)}to{-webkit-filter:blur(0);filter:blur(0);transform:none}}@keyframes blurInLeft{from{-webkit-filter:blur(15px);filter:blur(15px);transform:translate3d(-20%,0,0)}to{-webkit-filter:blur(0);filter:blur(0);transform:none}}@keyframes blurInRight{from{-webkit-filter:blur(15px);filter:blur(15px);transform:translate3d(20%,0,0)}to{-webkit-filter:blur(0);filter:blur(0);transform:none}}@keyframes blurInDown{from{-webkit-filter:blur(15px);filter:blur(15px);transform:translate3d(0,-20%,0)}to{-webkit-filter:blur(0);filter:blur(0);transform:none}}@keyframes fillIn{from{-webkit-filter:grayscale(100%);filter:grayscale(100%)}to{-webkit-filter:grayscale(0);filter:grayscale(0)}}', ANIMATION = 'fadeIn', ANIMATION_DURATION = '700ms', EASING = 'ease-in-out';

export { LOADER, LOADER_STYLE, PLACEHOLDER, PLACEHOLDER_STYLE, ANIMATIONS, ANIMATION, ANIMATION_DURATION, EASING };