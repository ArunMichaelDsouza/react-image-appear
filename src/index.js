import React, { Component } from 'React';

class ReactImageAppear extends Component {
    constructor(props) {
        super(props);

        this.onLoad = this.onLoad.bind(this);
    }

    onLoad() {
        console.log('loaded');
    }

    render() {
        return <img onLoad={this.onLoad} />
    }
}

export default ReactImageAppear;