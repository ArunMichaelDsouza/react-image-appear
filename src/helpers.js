// Helper functions

import { ANIMATIONS } from './constants';

const injectAnimationsScript = () => {
    let css = ANIMATIONS,
        head = document.head || document.getElementsByTagName('head')[0],
        style = document.createElement('style');

    style.type = 'text/css';
    if (style.styleSheet) {
        style.styleSheet.cssText = css;
    } else {
        style.appendChild(document.createTextNode(css));
    }

    head.appendChild(style);
};

export { injectAnimationsScript };