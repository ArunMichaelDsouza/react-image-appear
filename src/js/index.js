/*
    react-image-appear v1.0.0
    Copyright (c) 2018 Arun Michael Dsouza (amdsouza92@gmail.com)
    Licence: MIT
*/

import React, { Component } from 'React';
import PropTypes from 'prop-types';
import { LOADER, PLACEHOLDER_STYLES, ANIMATION, ANIMATION_DURATION, EASING } from './constants';

class ReactImageAppear extends Component {
    constructor(props) {
        super(props);

        this.state = {
            imgComponent: null
        };
        this.imageOnLoad = this.imageOnLoad.bind(this);
        this.getImageDimensions = this.getImageDimensions.bind(this);
        this.parseComputedDimensions = this.parseComputedDimensions.bind(this);
        this.createPlaceholder = this.createPlaceholder.bind(this);
    }

    componentDidMount() {
        const { src } = this.props;

        let imgElement;
        this.setState((prevState, props) => {
            return {
                imgComponent: React.createElement('img', {
                    src,
                    onLoad: this.imageOnLoad,
                    style: {
                        opacity: 0
                    },
                    ref: ref => {
                        imgElement = ref;
                    }
                })
            };
        }, () => {
            this.getImageDimensions(imgElement);
        });
    }

    imageOnLoad() {
        const { animation, animationDuration, easing } = this.props;

        console.log('loaded');
        this.setState((prevState, props) => {
            return {
                imgComponent: React.createElement('img', {
                    src: this.props.src,
                    style: {
                        animation: `${animation} ${animationDuration} ${easing}`
                    }
                })
            };
        });
    }

    getImageDimensions(imgElement) {
        const that = this,
            dimensionsInterval = setInterval(() => {
                const { width, height } = this.parseComputedDimensions(imgElement);

                if (width && height) {
                    clearInterval(dimensionsInterval);
                    that.createPlaceholder(width, height);
                }
            }, 10);
    }

    parseComputedDimensions(el) {
        return {
            width: Number(window.getComputedStyle(el).width.match(/\d+/)),
            height: Number(window.getComputedStyle(el).height.match(/\d+/))
        };
    }

    createPlaceholder(width, height) {
        const { imgComponent } = this.state,
            { loader, showLoader, placeholderStyles } = this.props,
            placeholderStyling = Object.assign({}, PLACEHOLDER_STYLES, placeholderStyles);

        const placeholder = React.createElement('div', {
            style: Object.assign({}, placeholderStyling, {
                width,
                height,
                backgroundImage: showLoader ? `url(${loader})` : null
            })
        }, React.cloneElement(imgComponent));

        this.setState((prevState, props) => {
            return {
                imgComponent: placeholder
            };
        });
    }

    render() {
        const { imgComponent } = this.state;

        return imgComponent;
    }
}

ReactImageAppear.propTypes = {
    src: PropTypes.string.isRequired,
    loader: PropTypes.string,
    animation: PropTypes.string,
    animationDuration: PropTypes.number,
    easing: PropTypes.string,
    showLoader: PropTypes.bool,
    placeholderStyles: PropTypes.object
};

ReactImageAppear.defaultProps = {
    loader: LOADER,
    animation: ANIMATION,
    animationDuration: ANIMATION_DURATION,
    easing: EASING,
    showLoader: true,
    placeholderStyles: {}
}

export default ReactImageAppear;