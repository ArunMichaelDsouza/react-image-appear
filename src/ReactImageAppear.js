/*
    react-image-appear v1.3.26
    Copyright (c) 2018 Arun Michael Dsouza (amdsouza92@gmail.com)
    Licence: MIT
*/

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
    LOADER, LOADER_STYLE, PLACEHOLDER, PLACEHOLDER_STYLE, ANIMATION, ANIMATION_DURATION, EASING
} from './constants';
import { injectAnimationsScript } from './helpers';

injectAnimationsScript();

class ReactImageAppear extends Component {
    constructor(props) {
        super(props);

        this.state = {
            imgComponent: null,
            loading: false
        };
        this.imageOnLoad = this.imageOnLoad.bind(this);
        this.getImageDimensions = this.getImageDimensions.bind(this);
        this.parseComputedDimensions = this.parseComputedDimensions.bind(this);
        this.createPlaceholder = this.createPlaceholder.bind(this);
        this.getPlaceholderStyle = this.getPlaceholderStyle.bind(this);
        this.createLoader = this.createLoader.bind(this);
        this.getLoaderStyle = this.getLoaderStyle.bind(this);
    }

    componentDidMount() {
        const { src, className } = this.props;

        let imgElement;
        this.setState((prevState, props) => {
            return {
                loading: true,
                imgComponent: React.createElement('img', {
                    src,
                    onLoad: this.imageOnLoad,
                    style: {
                        opacity: 0
                    },
                    className,
                    ref: ref => {
                        imgElement = ref;
                    }
                })
            };
        }, () => {
            this.getImageDimensions(imgElement);
        });
    }

    imageOnLoad() {
        const { src, animation, animationDuration, easing, className, onClick } = this.props;

        this.setState((prevState, props) => {
            return {
                loading: false,
                imgComponent: React.createElement('img', {
                    src,
                    style: {
                        animation: `${animation} ${animationDuration} ${easing}`
                    },
                    className,
                    onClick
                })
            };
        });
    }

    getImageDimensions(imgElement) {
        const that = this,
            dimensionsInterval = setInterval(() => {
                const { width, height } = this.parseComputedDimensions(imgElement);

                if (width && height) {
                    clearInterval(dimensionsInterval);
                    that.createPlaceholder(width, height);
                }
            }, 10);
    }

    parseComputedDimensions(el) {
        const widthMatch = window.getComputedStyle(el).width.match(/\d+\.?\d+/g);
        const heightMatch = window.getComputedStyle(el).height.match(/\d+\.?\d+/g);
        return {
            width: widthMatch ? Number(widthMatch.join('')) : undefined,
            height: heightMatch ? Number(heightMatch.join('')) : undefined,
        };
    }

    createPlaceholder(width, height) {
        const { imgComponent } = this.state,
            { placeholderClass } = this.props;

        const placeholder = React.createElement('div', {
            style: this.getPlaceholderStyle(width, height),
            className: placeholderClass
        }, React.cloneElement(imgComponent), this.createLoader());

        this.setState((prevState, props) => {
            return {
                imgComponent: placeholder
            };
        });
    }

    getPlaceholderStyle(width, height) {
        const { placeholderStyle, placeholder } = this.props;
        let placeholderStyling = Object.assign({}, PLACEHOLDER_STYLE, placeholderStyle);

        switch (typeof (placeholder)) {
            case 'boolean':
                if (placeholder) {
                    placeholderStyling = Object.assign({}, placeholderStyling, {
                        backgroundImage: `url(${PLACEHOLDER})`
                    })
                }
                break;
            case 'string':
                placeholderStyling = Object.assign({}, placeholderStyling, {
                    backgroundImage: `url(${placeholder})`
                })
                break;
        }
        return Object.assign({}, placeholderStyling, { width, height });
    }

    createLoader() {
        const { loader, loaderClass, showLoader } = this.props;
        const { loading } = this.state
        return (showLoader && loading ) ? React.createElement('img', {
            src: loader ? loader : LOADER,
            style: this.getLoaderStyle(),
            className: loaderClass
        }) : null;
    }

    getLoaderStyle() {
        const { loaderStyle } = this.props;

        return Object.assign({}, LOADER_STYLE, loaderStyle);
    }

    render() {
        const { imgComponent } = this.state;

        return imgComponent;
    }
}

ReactImageAppear.propTypes = {
    src: PropTypes.string.isRequired,
    loader: PropTypes.string,
    loaderStyle: PropTypes.object,
    loaderClass: PropTypes.string,
    placeholder: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.string
    ]),
    placeholderStyle: PropTypes.object,
    placeholderClass: PropTypes.string,
    animation: PropTypes.string,
    animationDuration: PropTypes.string,
    easing: PropTypes.string,
    showLoader: PropTypes.bool,
    className: PropTypes.string,
    onClick: PropTypes.func
};

ReactImageAppear.defaultProps = {
    loader: LOADER,
    loaderStyle: {},
    loaderClass: '',
    placeholder: false,
    placeholderStyle: {},
    placeholderClass: '',
    animation: ANIMATION,
    animationDuration: ANIMATION_DURATION,
    easing: EASING,
    showLoader: true,
    className: '',
    onClick: () => {}
};

export default ReactImageAppear;
