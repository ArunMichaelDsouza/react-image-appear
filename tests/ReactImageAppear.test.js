// Test for ReactImageAppear component

import React from 'react';
import { shallow, mount } from 'enzyme';
import ReactImageAppear from '../src/ReactImageAppear';
import { LOADER } from '../src/constants';
import renderer from 'react-test-renderer';

const src = 'https://newevolutiondesigns.com/images/freebies/tropical-beach-background-8.jpg';
const className = 'my-class';

describe('ReactImageAppear Component', () => {
    it('Snapshot test', () => {
        const tree = renderer
            .create(<ReactImageAppear src={src} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Check for "src" prop to be present and to have a string value', () => {
        const el = shallow(<ReactImageAppear src={src} />);

        expect(el.props().src).toEqual(src);
        expect(typeof el.props().src).toEqual('string');
    });

    it('Check for "className" prop to be present and to have a string value', () => {
        const el = shallow(<ReactImageAppear src={src} className={className} />);

        expect(el.props().className).toEqual(className);
        expect(typeof el.props().className).toEqual('string');
    });

    it('Check for "loader" prop to have a string value', () => {
        const el = mount(<ReactImageAppear src={src} />);

        expect(el.props().loader).toEqual(LOADER);
        expect(typeof el.props().loader).toEqual('string');
    });

    it('Check for "loader" prop to be custom and have a string value', () => {
        const loader = "https://cache.dominos.com/nolo/ca/en/010048/assets/build/images/img/spinner.gif",
            el = mount(<ReactImageAppear src={src} loader={loader} />);

        expect(el.props().loader).toEqual(loader);
        expect(typeof el.props().loader).toEqual('string');
    });

    it('Check for "loaderStyle" prop to have an object value', () => {
        const loaderStyle = { border: "2px solid red" },
            el = mount(<ReactImageAppear src={src} loaderStyle={loaderStyle} />);

        expect(el.props().loaderStyle).toEqual(loaderStyle);
        expect(typeof el.props().loaderStyle).toEqual('object');
    });

    it('Check for "loaderClass" prop to have a string value', () => {
        const loaderClass = 'my-loader',
            el = mount(<ReactImageAppear src={src} loaderClass={loaderClass} />);

        expect(el.props().loaderClass).toEqual(loaderClass);
        expect(typeof el.props().loaderClass).toEqual('string');
    });

    it('Check for "placeholder" prop to have a boolean value', () => {
        const el = mount(<ReactImageAppear src={src} placeholder />);

        expect(el.props().placeholder).toEqual(true);
        expect(typeof el.props().placeholder).toEqual('boolean');
    });

    it('Check for "placeholder" prop to be custom and to have a string value', () => {
        const placeholder = "http://getuikit.com/docs/images/placeholder_600x400.svg",
            el = mount(<ReactImageAppear src={src} placeholder={placeholder} />);

        expect(el.props().placeholder).toEqual(placeholder);
        expect(typeof el.props().placeholder).toEqual('string');
    });

    it('Check for "placeholderStyle" prop to have an object value', () => {
        const placeholderStyle = { border: "2px solid red", backgroundColor: 'black' },
            el = mount(<ReactImageAppear src={src} placeholderStyle={placeholderStyle} />);

        expect(el.props().placeholderStyle).toEqual(placeholderStyle);
        expect(typeof el.props().placeholderStyle).toEqual('object');
    });

    it('Check for "placeholderClass" prop to have a string value', () => {
        const placeholderClass = 'my-placeholder',
            el = mount(<ReactImageAppear src={src} placeholderClass={placeholderClass} />);

        expect(el.props().placeholderClass).toEqual(placeholderClass);
        expect(typeof el.props().placeholderClass).toEqual('string');
    });

    it('Check for "animation" prop to have a string value', () => {
        const animation = 'bounceIn',
            el = mount(<ReactImageAppear src={src} animation={animation} />);

        expect(el.props().animation).toEqual(animation);
        expect(typeof el.props().loader).toEqual('string');
    });

    it('Check for "animationDuration" prop to have a string value', () => {
        const animationDuration = '1s',
            el = mount(<ReactImageAppear src={src} animationDuration={animationDuration} />);

        expect(el.props().animationDuration).toEqual(animationDuration);
        expect(typeof el.props().animationDuration).toEqual('string');
    });

    it('Check for "easing" prop to have a string value', () => {
        const easing = 'ease-in',
            el = mount(<ReactImageAppear src={src} easing={easing} />);

        expect(el.props().easing).toEqual(easing);
        expect(typeof el.props().easing).toEqual('string');
    });

    it('Check for "showLoader" prop to have a boolean value', () => {
        const showLoader = false,
            el = mount(<ReactImageAppear src={src} showLoader={showLoader} />);

        expect(el.props().showLoader).toEqual(showLoader);
        expect(typeof el.props().showLoader).toEqual('boolean');
    });

    it('Check for rendered element to be an "img" with the expected src', () => {
        const el = mount(<ReactImageAppear src={src} />);

        expect(el.children().at(0).type()).toEqual('img');
        expect(el.find('img').prop('src')).toEqual(src);
    });

    it('Check for show and hide loading spinner', () => {
        const wrapper = shallow(<ReactImageAppear src={src} showLoader={true}/>);
        const componentInstance = wrapper.instance();
        componentInstance.componentDidMount();
        expect(wrapper.state('loading')).toBe(true);

        componentInstance.imageOnLoad();
        expect(wrapper.state('loading')).toBe(false);
    })
    
    it('Check for "onClick" prop to have a function value', () => {
        const onClick = () => {},
            el = mount(<ReactImageAppear src={src} onClick={onClick} />);

        expect(el.props().onClick).toEqual(onClick);
        expect(typeof el.props().onClick).toEqual('function');
    });
});
