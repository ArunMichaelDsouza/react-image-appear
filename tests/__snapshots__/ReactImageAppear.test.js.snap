// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactImageAppear Component Snapshot test 1`] = `
<img
  className=""
  onLoad={[Function]}
  src="https://newevolutiondesigns.com/images/freebies/tropical-beach-background-8.jpg"
  style={
    Object {
      "opacity": 0,
    }
  }
/>
`;
